### ISSUE TITLE
Method calls to non-existing methods of nested classes are resolved to outer class

### ISSUE THREAD

**Issue by oxisto** [javaparser/javaparser#2495]
Method calls to methods of nested classes are wrongly resolved to methods of the outer class, if the method does not exist on the nested class.

Woah! Explaining time. See the following example code can be used to re-produce the phenomena:

```java
public class OuterClass {

  public void test() {
    System.out.println("Test from OuterClass");
  }

  public void onlyOuter() {
    System.out.println("Only available on the outer");
  }

  public class NestedClass {

    public void test() {
      System.out.println("Test from NestedClass");
    }

  }

  protected void foo() {
    NestedClass n = new NestedClass();
    test(); // method from OuterClass
    n.test(); // method from NestedClass

    onlyOuter(); // method from OuterClass
    n.onlyOuter(); // not valid!
  }
}
```

- The `test()` call refers to `test` of the `OuterClass` and is correctly resolved
- The `n.test()` call refers to `test` of the `NestedClass` (or more correctly `OuterClass.NestedClass`) and is also correctly resolved
- The `onlyOuter()` is the same as `test()`, just to make the test more illustrative.

Now with `n.onlyOuter()` things go wrong. This would be a reference to a method `onlyOuter` of `NestedClass` (or its ancestral classes) and since it does not exist it should throw an exception that it cannot be resolved.

However, it is resolved as the `onlyOuter` function of `OuterClass`. The culprit seem to be `JavaParserTypeDeclarationAdapter#solveMethod` which somehow calls `context.getParent().solveMethod` if it cannot find any `candidateMethods` in the class itself. I guess the idea is to take scope into account but it seems to go wrong here.

**Comment by jlerbsc** [comment]
Hi @oxisto , @maartenc @MysterAitch it seems that the use case provided has compilation error so it makes sense that we have an operational error. I'm not sure this issue is relevant.

    "The method onlyOuter() is undefined for the type OuterClass.NestedClass"

**Comment by oxisto** [comment]
Hi @jlerbsc, yes the use case has an intentional error. Back in January, when you parsed the code with JavaParser you did not get an operational error, but rather the parse continued and associated a wrong method to it. So it could be that this is issue was fixed by some other changes in the mean time.

**Comment by jlerbsc** [comment]
To my knowledge the expected behavior is that the code must compile before you can you JP/JSS. It is not the case in your sample. If you think that the issue is solved you can close this issue or give us a simple unit test from this one.

    public class Issue2495Test extends AbstractSymbolResolutionTest {

    @Test
    public void test() {
        
        TypeSolver typeSolver = new ReflectionTypeSolver(false);
        ParserConfiguration config = new ParserConfiguration()
                .setSymbolResolver(new JavaSymbolSolver(typeSolver));
        StaticJavaParser.setConfiguration(config);

        String s =
                "public class OuterClass {\n" + 
                "\n" + 
                "  public void test() {\n" + 
                "  }\n" + 
                "\n" + 
                "  public void onlyOuter() {\n" + 
                "  }\n" + 
                "\n" + 
                "  public class NestedClass {\n" + 
                "\n" + 
                "    public void test() {\n" + 
                "    }\n" + 
                "\n" + 
                "  }\n" + 
                "\n" + 
                "  protected void foo() {\n" + 
                "    NestedClass n = new NestedClass();\n" + 
                "    n.onlyOuter(); // not valid!\n" + 
                "  }\n" + 
                "}";
        CompilationUnit cu = StaticJavaParser.parse(s);
        List<MethodCallExpr> mces = cu.findAll(MethodCallExpr.class);
        mces.forEach(mce -> {
            System.out.println(mce.toString());
            System.out.println(mce.resolve().getQualifiedSignature());
        });
        
    }
    }

**Comment by MysterAitch** [comment]
> To my knowledge the expected behavior is that the code must compile before you can you JP/JSS. 

Yes, this is generally the case!

Some exceptions do occur -- the grammar / parser is permissive in some places, and does allow some code that is syntactically valid but doesn't compile. This is generally picked up later in the language level checking stage. 

As a result, it is possible to make use of `LanguageLevel.RAW` to take advantage of this if you're only wanting to parse and possibly modify source code, but symbol resolution (type solving) almost certainly doesn't work with non-compiling code.

tldr: Cases where JavaParser works with non-compiling code should be considered as "happy accidents" :smile: 

_edit: not that this helps directly with this issue, of course! It sounds like this has been fixed since @oxisto first posted this issue. As @jlerbsc mentions, if there's still a problem let us know and we'll take another closer look! Alternatively, if you're happy with the current behaviour, then we can close this and the draft PR #2621 :+1:_

**Comment by oxisto** [comment]
Closing this. 

Unrelated to the issue but related to what  @MysterAitch  has said: There are some use cases (including the one we use JP for) where it makes sense to only analyse the syntactic structure of a program, since from a pure AST point of view the program is valid. We for example, are analysing just snippets of code where you possibly have no idea what kind of class the object you are looking at has. We do have some extra logic on top of JP that tries to solve that.

Until now my understanding was that this scenario is perfectly supported by JP and the user should just not expect the type solving / symbol solving to work, but at least as no syntax errors are in the code, i get the JP to a state where I can actually analyse things (which then might not be too accurate). So I would wish to avoid situations where JP actually just "exists" because a symbol is not resolved. As it was stated in the past, JP is not a compiler, but an analysis framework.
